File: tobas_fake_publishers/src/tobas_fake_publishers/fake_battery_publisher.py
import rclpy
from rclpy.duration import Duration

from tobas_tools_py.constants import Topic
from tobas_msgs.msg import Battery


class FakeBatteryPublisher:
    DEFAULT_SAMPLING_PERIOD = 0.01  # [m]
    DEFAULT_VOLTAGE = 14.8  # [V]
    DEFAULT_CURRENT = 20  # [A]

    def __init__(self) -> None:
        sampling_period = rclpy.get_param("~sampling_period", self.DEFAULT_SAMPLING_PERIOD)
        voltage = rclpy.get_param("~voltage", self.DEFAULT_VOLTAGE)
        current = rclpy.get_param("~current", self.DEFAULT_CURRENT)

        self._battery_msg = Battery()
        self._battery_msg.voltage = voltage
        self._battery_msg.current = current

        self._battery_pub = rclpy.Publisher(Topic.BATTERY, Battery, queue_size=1)
        self._timer = rclpy.Timer(Duration(sampling_period), self._timer_cb)

    def _timer_cb(self, event: rclpy.timer.TimerEvent) -> None:
        self._battery_msg.header.stamp = event.current_real
        self._battery_pub.publish(self._battery_msg)

File: tobas_fake_publishers/nodes/fake_battery_publisher.cpp
#include <tobas_node/node.hpp>
#include <tobas_constants/constants.hpp>
#include <tobas_msgs/msg/battery.hpp>

using namespace std;

class FakeBatteryPublisherNode : public tobas::BaseNode
{
  static constexpr auto kSamplingPeriod = 10ms;

  static constexpr double kDefaultVoltage = 14.8;  // [V]
  static constexpr double kDefaultCurrent = 20.;   // [A]

  using self = FakeBatteryPublisherNode;
  using super = tobas::BaseNode;

public:
  explicit FakeBatteryPublisherNode(const rclcpp::NodeOptions& options = rclcpp::NodeOptions());

private:
  double voltage_;
  double current_;

  ros2::PublisherPtr<tobas_msgs::msg::Battery> batt_pub_;
  ros2::TimerPtr timer_;

  void timerCb();
};

FakeBatteryPublisherNode::FakeBatteryPublisherNode(const rclcpp::NodeOptions& options)
  : super("fake_battery_publisher", options)
{
  voltage_ = getDoubleParam("voltage", kDefaultVoltage);
  current_ = getDoubleParam("current", kDefaultCurrent);

  batt_pub_ = createPublisher<tobas_msgs::msg::Battery>(tobas::kBatteryTopic);
  timer_ = createTimer(kSamplingPeriod, &self::timerCb, this);
}

void FakeBatteryPublisherNode::timerCb()
{
  auto batt_msg = std::make_unique<tobas_msgs::msg::Battery>();
  batt_msg->header.stamp = get_clock()->now();
  batt_msg->voltage = voltage_;
  batt_msg->current = current_;

  batt_pub_->publish(move(batt_msg));
}

RCLCPP_COMPONENTS_REGISTER_NODE(FakeBatteryPublisherNode)
